packages/apollo/src/ws.ts

```diff
@@ -0,0 +1,52 @@
+// * See https://github.com/enisdenjo/graphql-ws#graceful-restart
+import { Client, ClientOptions, createClient } from 'graphql-ws'
+
+export interface RestartableClient extends Client {
+  restart(): void
+  started(): boolean
+}
+
+export function createRestartableClient(options: ClientOptions): RestartableClient {
+  let restartRequested = false
+  let restart = () => {
+    restartRequested = true
+  }
+  let _started = false
+  const started = () => _started
+
+  const client = createClient({
+    ...options,
+    on: {
+      ...options.on,
+      connected: () => {
+        _started = true
+      },
+      opened: (socket: any) => {
```

Comment:
I would love to see the typing here. Even though we can't directly assign `WebSocket` to the function parameter because of how the `graphql-ws` typing is created, we can create a local variable and assign the type to it.

I was thinking about this approach:
```js
opened: (originalSocket) => { // it has a type of "unknown"
  const socket = originalSocket as WebSocket
  
  // rest of the code
}
```

From `graphql-ws`'s `client.d.ts` file:
> The first argument is actually the `WebSocket`, but to avoid
> bundling DOM typings because the client can run in Node env too,
> you should assert the websocket type during implementation.